File: C++/20.cpp
class Solution {
public:
    bool isValid(string s) {
        '''
        :type: s: str
        :rtype: bool
        '''
        map<char, char> par_map = {{'(', ')'}, {'{', '}'}, {'[', ']'}};
        stack<char> stk; 
        for (int i = 0; i < s.length(); i++) {
            if (s[i] == '(' || s[i] == '{' || s[i] == '[') 
                stk.push(s[i]);
            else { // close par
                if (stk.empty()) // no corresponding par 
                    return false;
                else {
                    if (s[i] != par_map[stk.top()])
                        return false;
                    else 
                        stk.pop();
                }
            }
        }
        return stk.empty();
    }
};

File: Python/20.py
class Solution:
    def isValid(self, s):
        '''
        type: s: str
        rtype: bool
        '''
        if len(s) % 2 == 1:
            return False
        d = {'(': ')', '{': '}', '[': ']'}
        stack = []
        for char in s:
            if char in d:   # '(', '{', '['
                stack.append(char)
            else:           # ')', '}', ']'
                if len(stack) == 0:
                    return False
                elif char != d[stack.pop()]:
                    return False
        return len(stack) == 0